elytron/WFCORE-4360-CredentialStore_Expression_Resolution.adoc

```diff
@@ -0,0 +1,210 @@
+
+NOTE: The credential store also offers a programatic API, should the tooling and operations be insufficient for an end user manual population of the credential store may remain an option.
+
+==== Expression Resolver
+
+Presently WildFly Core supports an expression resolver that can delegate to a Vault configuration and if that is not available fall back to use either system properties or environment variables.  This will be updated to make use of the `CapabilityRegistry` and attempt to lookup a capability using a pre-defined constant (To be defined) which exposes an expression resolver runtime API.  The expression resolver looked up using a capability will be used first to attempt to resolve any expression, if this is successful the resolved value will be used - if not we will fallback to the existing behaviour of using the vault and if that fails using a system property or environment variable.
+
+A new singleton resource will be added to the elytron subsystem called `expression=encryption` the purpose of this resource is to contain the configuration both for the encryption and decryption of values.  As a singleton we know only one instance of this resource can be defined within the subsystem reducing some of the complexity multiple instances would cause.
+
+The `expression=encryption` resource will register it's expression resolver capability with the `CapabilityRegistry` making it available for runtime resolution of expressions.
+
+The `expression=encryption` resource will contain a single attribute `resolvers` which will be used to define one or more resolvers to handle the encryption and decryption of values.  Each resolver will support the following configuration: -
+
+ * Reference to a credential store and alias to identify a specific previously created `SecretKey` to use.
+ * Name of the algorithm to use for the encryption / decryption.
+ * Any additional parameters such as `salt`, `iteration count` etc.. as required to configure the algorithm.
+
+The `expression=encryption` resource will also contain an operation `create-expression` the purpose of this operation being to take a clear text value and using one of the defined `resolvers` output an expression that can be used elsewhere in the management model.  The operation will just require two parameters: -
+
+ * `resolver` - The name of the defined resolver to use to encrypt the data.
+ * `data` - The clear text value to be encrypted.
+
+The return value of the operation will be the complete expression ready to be used elsewhere in the mode.
+
+The format of the expressions resolved using the WildFly Elytron expression resolver will be `${ENC:Resolver:ENCRYPTED_DATA}` where `Resolver` is a reference to the specific resolver defined in the `expression=encryption` resource and `ENCRYPTED_DATA` is the data to be decrypted encoded using Base64.
```

Comment:
Is there a reason why it is not simply ${CredentialStore:Name:Alias} instead of ${ENC:Resolver:ENCRYPTED_DATA}?

Reply:
The point here is we need to be cross referencing the new "expression=encryption" resource, when we reference the credential store directly we are only referencing a specific entry within that credential store but we still need to configure the Cipher used for the decryption.

Additionally within the credential store multiple credentials can be associated with a single alias so although today we are just looking to use a SecretKey in the future we may choose to use public / private keys etc..

Reply:
Hi @darranl, I'm looking to the RFE description and all linked issues. This RFE is about replacing vaults usage by CS. This is requested by users. AFAIK, vaults has this format: `${VAULT::VAULT_BLOCK_name::attribute_name::1}`. Originally, users have been used to use attribute name for vaults, not some base64 code. Readable attribute name is more user-friendly than some base-64 unreadable text.

> The point here is we need to be cross referencing the new "expression=encryption" resource, when we reference the credential store directly we are only referencing a specific entry within that credential store but we still need to configure the Cipher used for the decryption.

It sounds reasonable. So WDYT about this? Would this or similar approach be feasible in this situation? If not, why?
```
/subsystem=elytron/secret-key=custom-secret-key:add(generate-new=true)
/subsystem=elytron/credential-store=STORE_NAME:add(modifiable=true, implementation-properties={"cipher-type"="AES", cipher-parameters=[], "secret-key"=custom-secret-key}, create=true)
/subsystem=elytron/credential-store=STORE_NAME:add-alias(alias="Alias", secret-value="SENSITIVE_STRING")
"${CredentialStore:STORE_NAME:Alias}"
```

> Additionally within the credential store multiple credentials can be associated with a single alias so although today we are just looking to use a SecretKey in the future we may choose to use public / private keys etc..

These can be implemented by additional parameters to `/subsystem=elytron/secret-key=custom-secret-key:add(generate-new=true)` or can't it (why?)?

Thank you!

Reply:
Within this RFE we are not attempting to provide a direct 1:1 mapping with the Vault, the vault was used to handle the storage of the encrypted Strings that is not the change that is being made here.

One issue that we do have with RFEs is they are often expressed in terms of previously understood features and not in terms of the actual requirement to be solved.

Within this RFE the issue that we are specifically addressing is how to allow encryption of other attributes within the management model that are not credentials and the chosen approach is to support expressions containing inlined encrypted values which can be decrypted at runtime.

The Base64 part of this description is to specifically hide those details from an end user, those are an internal implementation detail.  The reason I am adding some additional internal details to those values is to give us options in the future if we want to change the representations / add additional functionality - in a future RFE we can add version 2 and make changes as needed.

The expression the end user will use will still be in a human readable format: -

  ${ENC:Resolver:ENCRYPTED_DATA}

Comparing this to the Vault solution the string 'ENC' is the equivalent of 'VAULT' from the prior expression. The value 'Resolver' allows the correct resolver to be identified and then the 'ENCRYPTED_DATA' is binary data so this needs to represented as a String so for this choosing to use Base64 but could equally use Hex.  So it is within this ENCRYPTED_DATA field only that some support for versioning is being added to add some future protection.

Regarding the 'expression=encryption' resource, one further option that could be considered is to make this a child resource of the credential-store resource.  Generally I am adamantly against child resources within the management model as they are often used to express a relationship where the parent depends on the child which is incorrect, however in this case the child would depend on the parent which would be correct.

If we did that however we have made extensive use of the capabilities and requirements APIs of the management tier of the server - this means other subsystems can advertise these capabilities and use them either with the elytron subsystem or as an alternative.  If we make the expression-encryption resource a child of the credential-store resource it will not be possible to use this with credential stores advertised by other subsystems, if however we make it an independent resource it can make use of credential stores advertised from elsewhere.

In both cases however we could also avoid using a singleton and instead use a named resource e.g.

expression-resolver=PREFIX

The expressions used in the management model could now be: -

  ${PREFIX:ENCRYPTED_DATA}

So now the PREFIX in the expression is used to cross reference the correct expressions-resolver.

Adding this level of configuration however does have the potential to make things a lot more complex, this RFE is bridging the gap between the core management of the server and subsystem provided capabilities - using a singleton drastically simplifies this.  If we switch to have multiple resources a lot more internal coordination of those resources becomes required.